### PR TITLE
Issue #2871403 by max-kuzomko: Add TextareaWidget and TextareaWidgetT…

### PR DESCRIPTION
**Pull request by Max-Kuzomko**
…est.

### CODE REVIEW

File: src/Plugin/TypedDataFormWidget/TextareaWidget.php [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      'resizable' => 'both',
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isApplicable(DataDefinitionInterface $definition) {
+    return is_subclass_of($definition->getClass(), StringInterface::class) ||
+      is_subclass_of($definition->getClass(), IntegerInterface::class) ||
```
Comment (fago) [inline src/Plugin/TypedDataFormWidget/TextareaWidget.php]: I don't think it makes sense to edit floats or integers in a textarea. Let's support strings only.

**Review by fago — CHANGES_REQUESTED** [review]
thx, that looks pretty solid!

File: tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  /**
+   * @covers ::isApplicable
+   */
+  public function testIsApplicable() {
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('any')));
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('binary')));
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('boolean')));
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('datetime_iso8601')));;
```
Comment (fago) [inline tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: I don't think this makes sense - a textarea for date strings does not fit very well. I think should be only be applicable to the 'string' data type for now.

**Review by fago — CHANGES_REQUESTED** [review]

File: tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    $this->widget = $this->getFormWidgetManager()->createInstance('textarea');
+  }
+
+  /**
+   * @covers ::isApplicable
+   */
+  public function testIsApplicable() {
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('string')));
+    $this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
+    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));
```
Comment (Max-Kuzomko) [inline tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: Thanks!
Fixed, but I am not sure should we leave the following code?
`$this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));`

Comment (fago) [reply tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: those two are fine, but the others should not be removed. There should be the whole list of data types in the test, so we can make sure the widget is available for the right types!

>
With
>I don't think this makes sense - a textarea for date strings does not fit very well. I think should be only be applicable to the 'string' data type for now.

I meant that we need to fix isApplicable() to return only true when it's a "string" data type.

**Review by fago — CHANGES_REQUESTED** [review]

File: tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php (hunk 3 above)
Comment (Max-Kuzomko) [reply tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: Sorry, I didn't get it.
Should we have the following testIsApplicable() ?

public function testIsApplicable() {
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('any')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('binary')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('boolean')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('datetime_iso8601')));;
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('duration_iso8601')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('email')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('float')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('integer')));
    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('timespan')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('timestamp')));
    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('uri')));
    $this->assertFalse($this->widget->isApplicable(ListDataDefinition::create('string')));
    $this->assertFalse($this->widget->isApplicable(MapDataDefinition::create()));
  }

Comment (fago) [reply tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: yes, exactly!

File: src/Plugin/TypedDataFormWidget/TextareaWidget.php (hunk 1 above)
Comment (fago) [reply src/Plugin/TypedDataFormWidget/TextareaWidget.php]: this still holds, so the code is wrong.

File: tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php (hunk 2 above)
Comment (fago) [reply tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: still the case.

File: tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+   */
+  public function testIsApplicable() {
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('any')));
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('binary')));
+    $this->assertFalse($this->widget->isApplicable(DataDefinition::create('boolean')));
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('datetime_iso8601')));;
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('duration_iso8601')));
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('email')));
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('float')));
+    $this->assertTrue($this->widget->isApplicable(DataDefinition::create('integer')));
```
Comment (fago) [inline tests/src/Functional/TypedDataFormWidget/TextareaWidgetTest.php]: still not as discussed.

**Review by fago — CHANGES_REQUESTED** [review]

**Pull request closed without merging**